File: backend/internal/database/seed.go
package database

import (
	"database/sql"
	"log"
	"time"

	"github.com/rs/xid"
	"golang.org/x/crypto/bcrypt"
)

func Seed(db *sql.DB) error {
	// Check if demo user already exists
	var count int
	err := db.QueryRow(`SELECT COUNT(*) FROM "User" WHERE email = ?`, "demo@trackside.com").Scan(&count)
	if err != nil {
		return err
	}
	if count > 0 {
		log.Println("Database already seeded, skipping...")
		return nil
	}

	log.Println("Seeding database...")
	now := time.Now().UTC().Format(time.RFC3339)

	// Create demo user
	passwordHash, err := bcrypt.GenerateFromPassword([]byte("password123"), 12)
	if err != nil {
		return err
	}

	userID := xid.New().String()
	_, err = db.Exec(`INSERT INTO "User" (id, name, email, passwordHash, experience, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?)`,
		userID, "Demo Driver", "demo@trackside.com", string(passwordHash), "INTERMEDIATE", now, now)
	if err != nil {
		return err
	}
	log.Println("Created demo user: demo@trackside.com")

	// Create car
	carID := xid.New().String()
	_, err = db.Exec(`INSERT INTO "Car" (id, make, model, year, userId, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?)`,
		carID, "Nissan", "350Z", 2006, userID, now, now)
	if err != nil {
		return err
	}

	// Create car mods
	mods := []struct{ name, category string }{
		{"BC Racing BR Coilovers", "SUSPENSION"},
		{"Tomei Expreme Ti Exhaust", "EXHAUST"},
		{"Z1 Motorsports Cold Air Intake", "ENGINE"},
		{"Stoptech ST-40 Big Brake Kit", "BRAKES"},
		{"Enkei RPF1 18x9.5", "WHEELS_TIRES"},
	}
	for _, m := range mods {
		_, err = db.Exec(`INSERT INTO "CarMod" (id, name, category, carId) VALUES (?, ?, ?, ?)`,
			xid.New().String(), m.name, m.category, carID)
		if err != nil {
			return err
		}
	}
	log.Println("Created demo car: 2006 Nissan 350Z with 5 mods")

	// Create track 1: Laguna Seca
	track1ID := xid.New().String()
	_, err = db.Exec(`INSERT INTO "Track" (id, name, location, state, latitude, longitude, description, uploadedById, status, isImported, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		track1ID, "Laguna Seca", "Monterey, CA", "CA", 36.5754, -121.7627,
		"Iconic road course featuring the famous Corkscrew turn. 2.238 miles, 11 turns, and significant elevation changes make this a challenging and rewarding track.",
		userID, "APPROVED", false, now, now)
	if err != nil {
		return err
	}

	// Track 1 events
	roadcourseEvent1ID := xid.New().String()
	_, err = db.Exec(`INSERT INTO "TrackEvent" (id, eventType, trackId) VALUES (?, ?, ?)`, roadcourseEvent1ID, "ROADCOURSE", track1ID)
	if err != nil {
		return err
	}
	_, err = db.Exec(`INSERT INTO "TrackEvent" (id, eventType, trackId) VALUES (?, ?, ?)`, xid.New().String(), "DRIFT", track1ID)
	if err != nil {
		return err
	}

	// Track 1 zones
	zones := []struct {
		name, desc string
		posX, posY float64
	}{
		{"The Corkscrew (T8-T8A)", "Famous downhill left-right combo with 5.5 stories of elevation change. Blind entry — use the tree as a braking marker.", 65, 25},
		{"Turn 2 (Andretti Hairpin)", "Tight left-hand hairpin. Late apex is key.", 30, 40},
		{"Turn 5", "High-speed left sweeper heading uphill. Carry momentum.", 45, 60},
	}

	zoneIDs := make([]string, len(zones))
	for i, z := range zones {
		zoneIDs[i] = xid.New().String()
		_, err = db.Exec(`INSERT INTO "TrackZone" (id, name, description, posX, posY, trackId, createdAt) VALUES (?, ?, ?, ?, ?, ?, ?)`,
			zoneIDs[i], z.name, z.desc, z.posX, z.posY, track1ID, now)
		if err != nil {
			return err
		}
	}
	log.Println("Created track: Laguna Seca")

	// Create track 2: Atlanta Motorsports Park
	track2ID := xid.New().String()
	_, err = db.Exec(`INSERT INTO "Track" (id, name, location, state, latitude, longitude, description, uploadedById, status, isImported, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		track2ID, "Atlanta Motorsports Park", "Dawsonville, GA", "GA", 34.3705, -84.1643,
		"A 2-mile, 16-turn road course with 100ft of elevation change, designed by Hermann Tilke. Features a dedicated drift pad and drag strip.",
		userID, "APPROVED", false, now, now)
	if err != nil {
		return err
	}

	// Track 2 events
	roadcourseEvent2ID := xid.New().String()
	_, err = db.Exec(`INSERT INTO "TrackEvent" (id, eventType, trackId) VALUES (?, ?, ?)`, roadcourseEvent2ID, "ROADCOURSE", track2ID)
	if err != nil {
		return err
	}
	_, err = db.Exec(`INSERT INTO "TrackEvent" (id, eventType, trackId) VALUES (?, ?, ?)`, xid.New().String(), "DRIFT", track2ID)
	if err != nil {
		return err
	}
	_, err = db.Exec(`INSERT INTO "TrackEvent" (id, eventType, trackId) VALUES (?, ?, ?)`, xid.New().String(), "DRAG", track2ID)
	if err != nil {
		return err
	}

	// Track 2 zones
	track2Zones := []struct {
		name, desc string
		posX, posY float64
	}{
		{"Turn 1", "Fast right-hander after the main straight. Heavy braking zone.", 80, 30},
		{"Turn 12 (Rollercoaster)", "Blind crest into a left-right combo. Commitment corner.", 35, 55},
	}
	for _, z := range track2Zones {
		_, err = db.Exec(`INSERT INTO "TrackZone" (id, name, description, posX, posY, trackId, createdAt) VALUES (?, ?, ?, ?, ?, ?, ?)`,
			xid.New().String(), z.name, z.desc, z.posX, z.posY, track2ID, now)
		if err != nil {
			return err
		}
	}
	log.Println("Created track: Atlanta Motorsports Park")

	// Zone tips
	tips := []struct {
		content string
		zoneIdx int
	}{
		{"Use the big tree on the left as your turn-in point. Trust the line and commit — hesitation here is dangerous.", 0},
		{"Brake deep and trail brake in. The car will rotate naturally. Get on power early for the uphill section.", 1},
		{"Stay wide and use all the road. The banking helps you carry more speed than you think.", 2},
	}
	for _, t := range tips {
		_, err = db.Exec(`INSERT INTO "ZoneTip" (id, content, conditions, zoneId, authorId, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?)`,
			xid.New().String(), t.content, "DRY", zoneIDs[t.zoneIdx], userID, now, now)
		if err != nil {
			return err
		}
	}
	log.Println("Created zone tips")

	// Review
	_, err = db.Exec(`INSERT INTO "TrackReview" (id, rating, content, conditions, trackId, trackEventId, authorId, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		xid.New().String(), 5,
		"Absolutely world-class track. The Corkscrew is every bit as intense as it looks on TV. The facility is well-maintained and the tech inspection is thorough. Will definitely be back.",
		"DRY", track1ID, roadcourseEvent1ID, userID, now, now)
	if err != nil {
		return err
	}
	log.Println("Created reviews")

	// Lap records
	_, err = db.Exec(`INSERT INTO "LapRecord" (id, lapTime, conditions, notes, tirePressureFL, tirePressureFR, tirePressureRL, tirePressureRR, fuelLevel, camberFL, camberFR, camberRL, camberRR, casterFL, casterFR, toeFL, toeFR, toeRL, toeRR, trackId, trackEventId, carId, driverId, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		xid.New().String(), "1:42.856", "DRY",
		"Best time of the day. Car felt great after adjusting front camber.",
		32.5, 32.5, 34.0, 34.0, 50.0, -2.5, -2.5, -1.8, -1.8, 5.2, 5.2, 0.1, 0.1, 0.15, 0.15,
		track2ID, roadcourseEvent2ID, carID, userID, now, now)
	if err != nil {
		return err
	}

	_, err = db.Exec(`INSERT INTO "LapRecord" (id, lapTime, conditions, notes, tirePressureFL, tirePressureFR, tirePressureRL, tirePressureRR, fuelLevel, camberFL, camberFR, camberRL, camberRR, trackId, trackEventId, carId, driverId, createdAt, updatedAt) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		xid.New().String(), "1:45.112", "WET",
		"Started raining mid-session. Dropped tire pressure to help with wet grip.",
		30.0, 30.0, 32.0, 32.0, 40.0, -2.5, -2.5, -1.8, -1.8,
		track2ID, roadcourseEvent2ID, carID, userID, now, now)
	if err != nil {
		return err
	}
	log.Println("Created lap records")

	log.Println("")
	log.Println("Seed complete! Login with:")
	log.Println("  Email: demo@trackside.com")
	log.Println("  Password: password123")
	return nil
}

File: seed.ts
import { PrismaClient } from '@prisma/client';
import bcrypt from 'bcryptjs';

const prisma = new PrismaClient();

async function main() {
    console.log('🌱 Seeding database...');

    // Create demo user
    const passwordHash = await bcrypt.hash('password123', 12);

    const user = await prisma.user.upsert({
        where: { email: 'demo@trackside.com' },
        update: {},
        create: {
            name: 'Demo Driver',
            email: 'demo@trackside.com',
            passwordHash,
            experience: 'INTERMEDIATE',
        },
    });

    console.log('✅ Created demo user:', user.email);

    // Create a car for the demo user
    const car = await prisma.car.create({
        data: {
            make: 'Nissan',
            model: '350Z',
            year: 2006,
            userId: user.id,
            mods: {
                create: [
                    { name: 'BC Racing BR Coilovers', category: 'SUSPENSION' },
                    { name: 'Tomei Expreme Ti Exhaust', category: 'EXHAUST' },
                    { name: 'Z1 Motorsports Cold Air Intake', category: 'ENGINE' },
                    { name: 'Stoptech ST-40 Big Brake Kit', category: 'BRAKES' },
                    { name: 'Enkei RPF1 18x9.5', category: 'WHEELS_TIRES' },
                ],
            },
        },
    });

    console.log('✅ Created demo car:', car.year, car.make, car.model);

    // Create sample tracks
    const track1 = await prisma.track.create({
        data: {
            name: 'Laguna Seca',
            location: 'Monterey, CA',
            state: 'CA',
            latitude: 36.5754,
            longitude: -121.7627,
            description: 'Iconic road course featuring the famous Corkscrew turn. 2.238 miles, 11 turns, and significant elevation changes make this a challenging and rewarding track.',
            uploadedById: user.id,
            events: {
                create: [
                    { eventType: 'ROADCOURSE' },
                    { eventType: 'DRIFT' },
                ],
            },
            zones: {
                create: [
                    {
                        name: 'The Corkscrew (T8-T8A)',
                        description: 'Famous downhill left-right combo with 5.5 stories of elevation change. Blind entry — use the tree as a braking marker.',
                        posX: 65,
                        posY: 25,
                    },
                    {
                        name: 'Turn 2 (Andretti Hairpin)',
                        description: 'Tight left-hand hairpin. Late apex is key.',
                        posX: 30,
                        posY: 40,
                    },
                    {
                        name: 'Turn 5',
                        description: 'High-speed left sweeper heading uphill. Carry momentum.',
                        posX: 45,
                        posY: 60,
                    },
                ],
            },
        },
        include: { events: true, zones: true },
    });

    console.log('✅ Created track:', track1.name);

    const track2 = await prisma.track.create({
        data: {
            name: 'Atlanta Motorsports Park',
            location: 'Dawsonville, GA',
            state: 'GA',
            latitude: 34.3705,
            longitude: -84.1643,
            description: 'A 2-mile, 16-turn road course with 100ft of elevation change, designed by Hermann Tilke. Features a dedicated drift pad and drag strip.',
            uploadedById: user.id,
            events: {
                create: [
                    { eventType: 'ROADCOURSE' },
                    { eventType: 'DRIFT' },
                    { eventType: 'DRAG' },
                ],
            },
            zones: {
                create: [
                    {
                        name: 'Turn 1',
                        description: 'Fast right-hander after the main straight. Heavy braking zone.',
                        posX: 80,
                        posY: 30,
                    },
                    {
                        name: 'Turn 12 (Rollercoaster)',
                        description: 'Blind crest into a left-right combo. Commitment corner.',
                        posX: 35,
                        posY: 55,
                    },
                ],
            },
        },
        include: { events: true, zones: true },
    });

    console.log('✅ Created track:', track2.name);

    // Create sample zone tips
    for (const zone of track1.zones) {
        await prisma.zoneTip.create({
            data: {
                content: zone.name.includes('Corkscrew')
                    ? 'Use the big tree on the left as your turn-in point. Trust the line and commit — hesitation here is dangerous.'
                    : zone.name.includes('Turn 2')
                        ? 'Brake deep and trail brake in. The car will rotate naturally. Get on power early for the uphill section.'
                        : 'Stay wide and use all the road. The banking helps you carry more speed than you think.',
                conditions: 'DRY',
                zoneId: zone.id,
                authorId: user.id,
            },
        });
    }

    console.log('✅ Created zone tips');

    // Create sample reviews
    const gripEvent1 = track1.events.find((e: any) => e.eventType === 'ROADCOURSE')!;
    await prisma.trackReview.create({
        data: {
            rating: 5,
            content: 'Absolutely world-class track. The Corkscrew is every bit as intense as it looks on TV. The facility is well-maintained and the tech inspection is thorough. Will definitely be back.',
            conditions: 'DRY',
            trackId: track1.id,
            trackEventId: gripEvent1.id,
            authorId: user.id,
        },
    });

    console.log('✅ Created reviews');

    // Create sample lap records
    const gripEvent2 = track2.events.find((e: any) => e.eventType === 'ROADCOURSE')!;
    await prisma.lapRecord.create({
        data: {
            lapTime: '1:42.856',
            conditions: 'DRY',
            notes: 'Best time of the day. Car felt great after adjusting front camber.',
            tirePressureFL: 32.5,
            tirePressureFR: 32.5,
            tirePressureRL: 34.0,
            tirePressureRR: 34.0,
            fuelLevel: 50,
            camberFL: -2.5,
            camberFR: -2.5,
            camberRL: -1.8,
            camberRR: -1.8,
            casterFL: 5.2,
            casterFR: 5.2,
            toeFL: 0.1,
            toeFR: 0.1,
            toeRL: 0.15,
            toeRR: 0.15,
            trackId: track2.id,
            trackEventId: gripEvent2.id,
            carId: car.id,
            driverId: user.id,
        },
    });

    await prisma.lapRecord.create({
        data: {
            lapTime: '1:45.112',
            conditions: 'WET',
            notes: 'Started raining mid-session. Dropped tire pressure to help with wet grip.',
            tirePressureFL: 30.0,
            tirePressureFR: 30.0,
            tirePressureRL: 32.0,
            tirePressureRR: 32.0,
            fuelLevel: 40,
            camberFL: -2.5,
            camberFR: -2.5,
            camberRL: -1.8,
            camberRR: -1.8,
            trackId: track2.id,
            trackEventId: gripEvent2.id,
            carId: car.id,
            driverId: user.id,
        },
    });

    console.log('✅ Created lap records');
    console.log('');
    console.log('🏁 Seed complete! Login with:');
    console.log('   Email: demo@trackside.com');
    console.log('   Password: password123');
}

main()
    .then(async () => {
        await prisma.$disconnect();
    })
    .catch(async (e) => {
        console.error(e);
        await prisma.$disconnect();
        process.exit(1);
    });
